### PR TITLE
Improve latency when one or more validators is unhealthy

### PR DESCRIPTION
**Pull request by mystenmark**
Yesterday, one of the validators was timing out requests, and it completely killed the node sync performance - 1/4th of requests were taking 8 seconds to complete.

This PR attempts to make the code mostly serial in the happy case (to not create undue load on validators) but becomes slightly parallel if requests do not complete quickly.

### CODE REVIEW

**Review by bmwill — APPROVED** [review]

File: crates/sui-core/src/authority_aggregator.rs [hunk 1]
```diff
@@ -633,29 +644,99 @@ where
...
+            // authority and then take the winner - this would create unnecessary load on
+            // authorities.
+            //
+            // The most efficient process from the network's point of view is to do one request at
+            // a time, however if the first validator that the client contacts is unavailable or
+            // slow, the client must wait for the serial_authority_request_timeout period to elapse
+            // before starting its next request.
+            //
+            // So, this process is designed as a compromise between these two extremes.
+            // - We start one request, and schedule another request to begin after
```
Comment (gdanezis) [inline crates/sui-core/src/authority_aggregator.rs:696]: It is still inefficient to timeout on every request, even if the timeout to contact a second authority is shorter than the full timeout. Have a look at the `AuthorityHealth` information that is stored within the `ActiveAuthority` and shared between all active services of a node. It allows the active services upon failure / timeout or byz behaviour to register the event as a minimum amount of time before contacting this authority again. This way upon the first failure to respond we can remember to not contact an authority again for some time, or prefer others.

Feel free to augment the Health information to suit your needs too.

**Review by gdanezis — COMMENTED** [review]
Nice one, but also try to remember the slowness / failure and de-prioritize this authority in subsequent requests?

File: crates/sui-core/src/authority_aggregator.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/authority_aggregator.rs:696]: Wow, wasn't aware of that.

The AuthorityHealth should really be used by every part of the AuthorityAggregator, but that's more than enough work for another PR. I will create an issue so we don't lose track of this item.

Comment (mystenmark) [reply crates/sui-core/src/authority_aggregator.rs:696]: Filed #3515

**Pull request merged**